torch/_inductor/fx_passes/bucketing.py

```diff
@@ -277,523 +225,364 @@ def _rank_idx_dict(group_name: str) -> dict[int, int]:
             continue
         collected_node_set.add(user_node)
         find_recursive_users_of_fx_node(
             user_node,
             collected_node_set,
-            criteria_cb=criteria_cb,
         )
 
 
+def reduce_scatter_merge_fn_to_trace(
+    rs_ins: list[torch.Tensor],
+    group_size: int,
+    group_name: str,
+    reduce_op: str,
+    reduce_dtype: torch.dtype,  # type: ignore[name-defined]
+    device: torch.device,  # type: ignore[name-defined]
+) -> list[torch.Tensor]:  # type: ignore[no-untyped-def]
+    rs_ins_flattened = [rs_in.view(-1) for rs_in in rs_ins]
+
+    rs_ins_srcs = [
+        rs_in_f.split([rs_in_f.numel() // group_size] * group_size)
+        for rs_in_f in rs_ins_flattened
+    ]
+
+    foreach_copy_srcs = []
+    for rank_idx in range(group_size):
+        for rs_in_idx in range(len(rs_ins)):
+            foreach_copy_srcs.append(rs_ins_srcs[rs_in_idx][rank_idx])
+
+    new_rs_in = torch.cat(foreach_copy_srcs, dim=0)
+
+    wait_tensor = torch.ops.c10d_functional.wait_tensor(
+        torch.ops._c10d_functional.reduce_scatter_tensor.default(
+            new_rs_in, reduce_op, group_size, group_name
+        )
+    )
+    new_rs_out = wait_tensor
+
+    new_outs = []
+    new_rs_out_offset = 0
+    for rs_in in rs_ins:
+        new_out_size = torch.Size((rs_in.shape[0] // group_size,) + rs_in.shape[1:])  # type: ignore[attr-defined]
+        new_out = new_rs_out.narrow(0, new_rs_out_offset, new_out_size.numel()).reshape(
+            new_out_size
+        )
+        new_outs.append(new_out)
+        new_rs_out_offset += new_out_size.numel()
+    return new_outs
+
+
+def all_gather_merge_fn_to_trace(
+    ag_ins: list[torch.Tensor],
+    group_size: int,
+    group_name: str,
+    dtype: torch.dtype,  # type: ignore[name-defined]
+    local_rank: int,
+) -> list[torch.Tensor]:
+    ins_sizes = [ag_in.shape for ag_in in ag_ins]
+    ins_split_sizes = [ag_in.numel() for ag_in in ag_ins]
+    ag_input_numel = sum(ins_split_sizes)
+    device = ag_ins[0].device
+    new_ag_out = torch.empty(ag_input_numel * group_size, dtype=dtype, device=device)
+    new_ag_in = new_ag_out.narrow(0, ag_input_numel * local_rank, ag_input_numel)
+    foreach_copy_dsts = torch.split(new_ag_in, ins_split_sizes)
+    ag_ins_flattened = [ag_in.reshape(-1) for ag_in in ag_ins]
+    torch._foreach_copy_(foreach_copy_dsts, ag_ins_flattened)
+    wait_tensor = torch.ops.c10d_functional.wait_tensor(
+        torch.ops._c10d_functional.all_gather_into_tensor_out.default(
+            new_ag_in, group_size, group_name, out=new_ag_out
+        )
+    )
+    new_ag_out_reshaped = wait_tensor.reshape(group_size, -1)
+    outs = torch.split_with_sizes(
+        new_ag_out_reshaped,
+        ins_split_sizes,
+        dim=1,
+    )
+    outs_reshaped = [
+        o.reshape((shape[0] * group_size,) + shape[1:])
+        for o, shape in zip(outs, ins_sizes)
+    ]
+    return outs_reshaped
+
+
+def all_gather_merge_fn_to_trace_fsdp_ag_copy_in(
```

Comment:
significant code duplication from `all_gather_merge_fn_to_trace` - consider adding a kwarg 'use_fsdp_copy_in` and branching on that?

Reply:
well i wrote that before i saw the '_functional' one. Idk, maybe its ok to keep separate.  or, merge the fsdp_copy_in one with the functional one via kwarg

Reply:
Ok, will merge fsdp_copy_in and _functional.